sql-plugin/src/main/scala/org/apache/spark/sql/rapids/urlFunctions.scala

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.rapids
+
+import java.net.URISyntaxException
+
+import ai.rapids.cudf.{ColumnVector, DType, RegexProgram, Scalar, Table}
+import com.nvidia.spark.rapids._
+import com.nvidia.spark.rapids.Arm._
+import com.nvidia.spark.rapids.RapidsPluginImplicits._
+import com.nvidia.spark.rapids.shims.ShimExpression
+
+import org.apache.spark.sql.catalyst.analysis.TypeCheckResult
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.internal.SQLConf
+import org.apache.spark.sql.rapids.shims.RapidsErrorUtils
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.unsafe.types.UTF8String
+
+object GpuParseUrl {
+  private val HOST = "HOST"
+  private val PATH = "PATH"
+  private val QUERY = "QUERY"
+  private val REF = "REF"
+  private val PROTOCOL = "PROTOCOL"
+  private val FILE = "FILE"
+  private val AUTHORITY = "AUTHORITY"
+  private val USERINFO = "USERINFO"
+  private val REGEXPREFIX = """(&|^|\?)"""
+  private val REGEXSUBFIX = "=([^&]*)"
+}
+
+case class GpuParseUrl(children: Seq[Expression], 
+    failOnErrorOverride: Boolean = SQLConf.get.ansiEnabled) 
+  extends GpuExpression with ShimExpression with ExpectsInputTypes {
+
+  def this(children: Seq[Expression]) = this(children, SQLConf.get.ansiEnabled)
+
+  override def nullable: Boolean = true
+  override def inputTypes: Seq[DataType] = Seq.fill(children.size)(StringType)
+  override def dataType: DataType = StringType
+  override def prettyName: String = "parse_url"
+
+  import GpuParseUrl._
+  
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.size > 3 || children.size < 2) {
+      RapidsErrorUtils.parseUrlWrongNumArgs(children.size) match {
+        case res: Some[TypeCheckResult] => return res.get
+        case _ => // error message has been thrown
+      }
+    }
+    super[ExpectsInputTypes].checkInputDataTypes()
+  }
+
+  private def getPattern(key: UTF8String): RegexProgram = {
+    val regex = REGEXPREFIX + key.toString + REGEXSUBFIX
```

Comment:
This brings up a kind of interesting corner case.  Technically the key is a regular expression. I could mess things up by putting in a .* in the name of the key and it would match. Our regular expressions are not bit for bit compatible with Sparks.  I am not sure if it is better for us to escape the regular expression and try to be sure that we treat it as a literal value, or if we try to transpile the resulting regular expression, or if we just document that this might be different.

I personally think that this is a bug so I filed https://issues.apache.org/jira/browse/SPARK-44500 for it. I would prefer that we fix the problem in our code and document that it is different.

Reply:
cuDF regex does not support \Q and \E, so we cannot use `Pattern.quote` directly here. 

We also can't add `\`s to escape all special characters, because the literal character set in Characters (any character except `[\^$.⎮?*+()`) and Character Classes (any character except `\^-]`) are different.

I suggest creating a follow-up issue on this if you think it is ok.